ODIN_II/SRC/HardSoftLogicMixer.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "HardSoftLogicMixer.hpp"
+
+#include <stdint.h> // INT_MAX
+#include <vector>
+
+#include "multipliers.h"       // instantiate_simple_soft_multiplier
+#include "netlist_statistic.h" // mixing_optimization_stats
+#include "odin_error.h"        // error_message
+
+HardSoftLogicMixer::HardSoftLogicMixer(const config_t& config) {
+    _allOptsDisabled = true;
+    // By default, disables all optimizations
+    parse_opt_parameters(config);
+}
+
+void HardSoftLogicMixer::parse_opt_parameters(const config_t config) {
+    if (config.mults_ratio != -1.0) {
+        MultsOpt* opt = new MultsOpt(config.mults_ratio);
+        this->_enabled_opts.push_back(opt);
+        _allOptsDisabled = false;
+    } else if (config.exact_mults != -1) {
+        MultsOpt* opt = new MultsOpt(config.exact_mults);
+        this->_enabled_opts.push_back(opt);
+        _allOptsDisabled = false;
+    }
+}
+
+void HardSoftLogicMixer::note_candidate_node(nnode_t* opNode, operation_list type) {
+    _nodes_by_opt[type].push_back(opNode);
+}
+
+bool HardSoftLogicMixer::softenable(operation_list type) {
+    bool result = false;
+    if (!this->_allOptsDisabled) {
+        for (MixingOpt* opt : _enabled_opts) {
```

Comment:
I think I would prefer having this be a fixed array. just access opt[type], and check that it has been specialized (aka it is not using the base class)

Reply:
Thank you, changed!